### PR TITLE
feature/better-amount-input

### PR DESCRIPTION
**Pull request by DataCourier**
Work in progress, seeking comments right now.
- [x] finish test suite
- [x] simplify code
- [x] add some comments
- [x] get feedback
- [x] apply component to the other form as well
- [x] branch needs to be renamed
- [x] squash commits and commit with nice messages
- [ ] <del>modify to support mili-satoshis</del>

feb-11-2018 14-44-55

### CODE REVIEW

File: app/components/Form/AmountInput.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  handleChange(e) {
+    let value = e.target.value;
+    let integer = null;
+    let fractional = null;
+
+    if (value.match(/[0-9]+/)) {
+      integer = e.target.value * 1;
+    }
+    if (value.match(/[0-9]*[,\.][0-9]+/)) {
+      [integer, fractional] = value.toString().split('.');
```
Comment (Empact) [inline app/components/Form/AmountInput.js]: The split could be a `.` or ` `,` as per the regex.

File: app/components/Form/AmountInput.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+  handleChange(e) {
+    let value = e.target.value;
+    let integer = null;
+    let fractional = null;
+
+    if (value.match(/[0-9]+/)) {
+      integer = e.target.value * 1;
+    }
+    if (value.match(/[0-9]*[,\.][0-9]+/)) {
```
Comment (Empact) [inline app/components/Form/AmountInput.js]: No need to escape the `.` inside a character class.

File: app/components/Form/AmountInput.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      [integer, fractional] = value.toString().split('.');
+    }
+
+    // limit fractional precision to 8 places
+    if (fractional && fractional.length > 8) {
```
Comment (Empact) [inline app/components/Form/AmountInput.js]: Limit is 11 characters for off-chain, 8 for on-chain.
https://github.com/lightningnetwork/lnd/blob/master/lnwire/msat.go

Comment (Empact) [reply app/components/Form/AmountInput.js]: This is incorrect - satoshis are still the unit for off-chain payments. Higher precision is used to accommodate fees.

File: app/components/Form/AmountInput.js (hunk 1 above)
Comment (DataCourier) [reply app/components/Form/AmountInput.js]: fixed

File: app/components/Form/AmountInput.js (hunk 2 above)
Comment (DataCourier) [reply app/components/Form/AmountInput.js]: fixed

File: app/components/Form/AmountInput.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import CurrencyIcon from 'components/CurrencyIcon'
+import styles from './RequestForm.scss'
+
+class AmountInput extends React.Component {
+
```
Comment (JimmyMow) [inline app/components/Form/AmountInput.js]: no need for empty line here

File: app/components/Form/AmountInput.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    // limit fractional precision to 8 places
+    if (fractional && fractional.length > 8) {
+      fractional = fractional.substring(0, 8)
+    }
+
+    return [integer, fractional]
+  }
+
+  formatValue(integer, fractional) {
+    let value;
```
Comment (JimmyMow) [inline app/components/Form/AmountInput.js]: our code style is no semicolons for now :p

File: app/components/Form/AmountInput.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+      </section>
+    )
+  }
+}
+
+AmountInput.propTypes = {
+  currency: PropTypes.string.isRequired,
+  crypto: PropTypes.string.isRequired,
+  onChange:PropTypes.func.isRequired,
+  amount: PropTypes.string.isRequired,
```
Comment (JimmyMow) [inline app/components/Form/AmountInput.js]: can remove the trailing comma

File: app/components/Form/AmountInput.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    let fractional = null
+
+    if (value == '') { value = '0' }
+    if (value*1.0 < 0) { value = '0.0' }
+
+    // simplify this, make flat if possible
+    // extract the parsing bit?
+    if (value.match(/^[0-9]*[.,][0-9]*$/)) {
+      [integer, fractional] = value.toString().split(/[.,]/)
+      if (!fractional || fractional == '') {
```
Comment (JimmyMow) [inline app/components/Form/AmountInput.js]: hmm, is this conditional needed? if `fractional` evaluates to false or `fractional == ''` then the conditional below will evaluate to false anyways right?

File: app/components/Form/AmountInput.js [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    return [integer, fractional]
+  }
+
+  formatValue(integer, fractional) {
+    let value;
+    if (fractional && fractional.length > 0) {
+      value = `${integer}.${fractional}`
+    } else { 
+      if (fractional === '') {
```
Comment (JimmyMow) [inline app/components/Form/AmountInput.js]: this conditional sets value = `${integer}.` no matter what. doesn't look like it's needed

File: app/components/Form/RequestForm.js [hunk 9]
```diff
@@ -2,6 +2,7 @@ import React from 'react'
 import PropTypes from 'prop-types'
 import CurrencyIcon from 'components/CurrencyIcon'
 import styles from './RequestForm.scss'
+import AmountInput from './AmountInput'
```
Comment (JimmyMow) [inline app/components/Form/RequestForm.js]: Should be imported above the `styles`

**Comment by JimmyMow** [comment]
Nice @DataCourier! This is an overall improvement. I left some small comments throughout, looks good overall. A few changes and passing travis and I'd love to merge this. 

Also, why did you not use this in the `PayForm`? Just curious if there was a specific reason

**Comment by JimmyMow** [comment]
Also extracting out the `AmountInput` into its own component is a wonderful idea and a step in the right direction. I'm going to be adding multi-currencies support soon (BTC, bits, USD, etc). This component will be super helpful then as this form gets more complex

**Comment by DataCourier** [comment]
@JimmyMow I'll be applying it in the other form as well (its in the checklist), just wanted to get general feedback on the behavior first. Will be addressing the comments over the weekend most likely.

File: app/components/Form/AmountInput.js (hunk 8 above)
Comment (DataCourier) [reply app/components/Form/AmountInput.js]: When integer is null, it runs the else part. Thats how it knows its just `X.` vs just `X`

**Comment by DataCourier** [comment]
Renamed branch, closing PR for the new one https://github.com/LN-Zap/zap-desktop/pull/322

**Pull request closed without merging**